File: bindings/kotlin/src/main/kotlin/at/schem/nucleation/StoreIo.kt
package at.schem.nucleation;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface StoreIoLib: Library {
    fun StoreIo_destroy(handle: Pointer)
    fun StoreIo_open(uri: Slice): ResultPointerInt
    fun StoreIo_save(schematic: Pointer, uri: Slice, version: Slice): ResultUnitInt
    fun StoreIo_export_settings_schema(format: Slice, write: Pointer): ResultUnitInt
    fun StoreIo_import_settings_schema(format: Slice, write: Pointer): ResultUnitInt
    fun StoreIo_supported_import_formats(write: Pointer): ResultUnitInt
    fun StoreIo_supported_export_formats(write: Pointer): ResultUnitInt
    fun StoreIo_format_versions(format: Slice, write: Pointer): ResultUnitInt
    fun StoreIo_default_format_version(format: Slice, write: Pointer): ResultUnitInt
}
/** Namespace type for the URI-based transparent I/O and format-manager
*queries (PORTING rule 12).
*/
class StoreIo internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
    internal var owned: Boolean,
)  {

    init {
        if (this.owned) {
            this.registerCleaner()
        }
    }

    private class StoreIoCleaner(val handle: Pointer, val lib: StoreIoLib) : Runnable {
        override fun run() {
            lib.StoreIo_destroy(handle)
        }
    }
    private fun registerCleaner() {
        CLEANER.register(this, StoreIo.StoreIoCleaner(handle, StoreIo.lib));
    }

    companion object {
        internal val libClass: Class<StoreIoLib> = StoreIoLib::class.java
        internal val lib: StoreIoLib = Native.load("nucleation", libClass)
        @JvmStatic
        
        /** Open a schematic from a URI: a local path, `file://...`, or
        *`s3://bucket/key.schem`. The format is auto-detected from the URI's
        *extension. Single-string URIs for `redis://`, `postgres://`, and
        *`mem://` are rejected by the core resolver; use `Store::open_schematic`
        *with an explicit store for those backends.
        */
        fun open(uri: String): Result<Schematic> {
            val uriSliceMemory = PrimitiveArrayTools.borrowUtf8(uri)
            
            val returnVal = lib.StoreIo_open(uriSliceMemory.slice);
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    val selfEdges: List<Any> = listOf()
                    val handle = nativeOkVal 
                    val returnOpaque = Schematic(handle, selfEdges, true)
                    return returnOpaque.ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                uriSliceMemory.close()
            }
        }
        @JvmStatic
        
        /** Save a schematic to a URI: a local path, `file://...`, or
        *`s3://bucket/key.schem`. The format is auto-detected from the URI's
        *extension; `version` selects the format version (empty string =
        *format default). Single-string URIs for `redis://`, `postgres://`, and
        *`mem://` are rejected by the core resolver; use `Store::save_schematic`
        *with an explicit store for those backends.
        */
        fun save(schematic: Schematic, uri: String, version: String): Result<Unit> {
            val uriSliceMemory = PrimitiveArrayTools.borrowUtf8(uri)
            val versionSliceMemory = PrimitiveArrayTools.borrowUtf8(version)
            
            val returnVal = lib.StoreIo_save(schematic.handle, uriSliceMemory.slice, versionSliceMemory.slice);
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    return Unit.ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                uriSliceMemory.close()
                versionSliceMemory.close()
            }
        }
        @JvmStatic
        
        /** The JSON schema describing the export settings of `format`. Errors
        *with `NotFound` for an unknown format.
        */
        fun exportSettingsSchema(format: String): Result<String> {
            val formatSliceMemory = PrimitiveArrayTools.borrowUtf8(format)
            val write = DW.lib.diplomat_buffer_write_create(0)
            val returnVal = lib.StoreIo_export_settings_schema(formatSliceMemory.slice, write);
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    
                    val returnString = DW.writeToString(write)
                    return returnString.ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                formatSliceMemory.close()
            }
        }
        @JvmStatic
        
        /** The JSON schema describing the import settings of `format`. Errors
        *with `NotFound` for an unknown format.
        */
        fun importSettingsSchema(format: String): Result<String> {
            val formatSliceMemory = PrimitiveArrayTools.borrowUtf8(format)
            val write = DW.lib.diplomat_buffer_write_create(0)
            val returnVal = lib.StoreIo_import_settings_schema(formatSliceMemory.slice, write);
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    
                    val returnString = DW.writeToString(write)
                    return returnString.ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                formatSliceMemory.close()
            }
        }
        @JvmStatic
        
        /** The supported import formats, written as a JSON array string.
        */
        fun supportedImportFormats(): Result<String> {
            val write = DW.lib.diplomat_buffer_write_create(0)
            val returnVal = lib.StoreIo_supported_import_formats(write);
            val nativeOkVal = returnVal.getNativeOk();
            if (nativeOkVal != null) {
                
                val returnString = DW.writeToString(write)
                return returnString.ok()
            } else {
                return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
            }
        }
        @JvmStatic
        
        /** The supported export formats, written as a JSON array string.
        */
        fun supportedExportFormats(): Result<String> {
            val write = DW.lib.diplomat_buffer_write_create(0)
            val returnVal = lib.StoreIo_supported_export_formats(write);
            val nativeOkVal = returnVal.getNativeOk();
            if (nativeOkVal != null) {
                
                val returnString = DW.writeToString(write)
                return returnString.ok()
            } else {
                return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
            }
        }
        @JvmStatic
        
        /** The known versions of an export format, written as a JSON array string
        *(empty array for an unknown format, matching the old ABI).
        */
        fun formatVersions(format: String): Result<String> {
            val formatSliceMemory = PrimitiveArrayTools.borrowUtf8(format)
            val write = DW.lib.diplomat_buffer_write_create(0)
            val returnVal = lib.StoreIo_format_versions(formatSliceMemory.slice, write);
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    
                    val returnString = DW.writeToString(write)
                    return returnString.ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                formatSliceMemory.close()
            }
        }
        @JvmStatic
        
        /** The default version of an export format. Errors with `NotFound` for an
        *unknown format.
        */
        fun defaultFormatVersion(format: String): Result<String> {
            val formatSliceMemory = PrimitiveArrayTools.borrowUtf8(format)
            val write = DW.lib.diplomat_buffer_write_create(0)
            val returnVal = lib.StoreIo_default_format_version(formatSliceMemory.slice, write);
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    
                    val returnString = DW.writeToString(write)
                    return returnString.ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                formatSliceMemory.close()
            }
        }
    }

}

File: bindings/php/StoreIo.php
<?php
namespace Stencil;

final class StoreIo {
    /** @internal */
    public \FFI\CData $ptr;
    private bool $owned;

    /** @internal */
    public function __construct(\FFI\CData $ptr, bool $owned) {
        $this->ptr = $ptr;
        $this->owned = $owned;
    }

    public static function open(string $uri) {
        $__n0 = strlen($uri);
        $__view0 = Lib::ffi()->new('DiplomatStringView');
        if ($__n0 > 0) {
            $__buf0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            \FFI::memcpy($__buf0, $uri, $__n0);
            $__view0->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf0[0]));
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $result = Lib::ffi()->StoreIo_open($__view0);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return new Schematic($result->ok, true);
    }

    public static function save( $schematic, string $uri, string $version) {
        $__n1 = strlen($uri);
        $__view1 = Lib::ffi()->new('DiplomatStringView');
        if ($__n1 > 0) {
            $__buf1 = Lib::ffi()->new("uint8_t[" . $__n1 . "]", false);
            \FFI::memcpy($__buf1, $uri, $__n1);
            $__view1->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf1[0]));
        } else {
            $__view1->data = null;
        }
        $__view1->len = $__n1;
        $__n2 = strlen($version);
        $__view2 = Lib::ffi()->new('DiplomatStringView');
        if ($__n2 > 0) {
            $__buf2 = Lib::ffi()->new("uint8_t[" . $__n2 . "]", false);
            \FFI::memcpy($__buf2, $version, $__n2);
            $__view2->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf2[0]));
        } else {
            $__view2->data = null;
        }
        $__view2->len = $__n2;
        $result = Lib::ffi()->StoreIo_save($schematic->ptr, $__view1, $__view2);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
    }

    public static function exportSettingsSchema(string $format) {
        $__n0 = strlen($format);
        $__view0 = Lib::ffi()->new('DiplomatStringView');
        if ($__n0 > 0) {
            $__buf0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            \FFI::memcpy($__buf0, $format, $__n0);
            $__view0->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf0[0]));
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        $result = Lib::ffi()->StoreIo_export_settings_schema($__view0, $write);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return Lib::readAndFreeWrite($write);
    }

    public static function importSettingsSchema(string $format) {
        $__n0 = strlen($format);
        $__view0 = Lib::ffi()->new('DiplomatStringView');
        if ($__n0 > 0) {
            $__buf0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            \FFI::memcpy($__buf0, $format, $__n0);
            $__view0->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf0[0]));
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        $result = Lib::ffi()->StoreIo_import_settings_schema($__view0, $write);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return Lib::readAndFreeWrite($write);
    }

    public static function supportedImportFormats() {
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        $result = Lib::ffi()->StoreIo_supported_import_formats($write);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return Lib::readAndFreeWrite($write);
    }

    public static function supportedExportFormats() {
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        $result = Lib::ffi()->StoreIo_supported_export_formats($write);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return Lib::readAndFreeWrite($write);
    }

    public static function formatVersions(string $format) {
        $__n0 = strlen($format);
        $__view0 = Lib::ffi()->new('DiplomatStringView');
        if ($__n0 > 0) {
            $__buf0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            \FFI::memcpy($__buf0, $format, $__n0);
            $__view0->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf0[0]));
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        $result = Lib::ffi()->StoreIo_format_versions($__view0, $write);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return Lib::readAndFreeWrite($write);
    }

    public static function defaultFormatVersion(string $format) {
        $__n0 = strlen($format);
        $__view0 = Lib::ffi()->new('DiplomatStringView');
        if ($__n0 > 0) {
            $__buf0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            \FFI::memcpy($__buf0, $format, $__n0);
            $__view0->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf0[0]));
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        $result = Lib::ffi()->StoreIo_default_format_version($__view0, $write);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return Lib::readAndFreeWrite($write);
    }

    public function __destruct() {
        if ($this->owned) {
            Lib::ffi()->StoreIo_destroy($this->ptr);
        }
    }
}
